s3cmd

```diff
@@ -583,12 +594,79 @@ def cmd_object_del(args):
 
         if not recursive:
             rc = subcmd_object_del_uri(uri_str)
-        else:
+        elif Config().exclude:
             rc = subcmd_batch_del(uri_str = uri_str)
+        else:
+            rc = subcmd_batch_del_iterative(uri_str = uri_str)
         if not rc:
             return rc
     return EX_OK
 
+def subcmd_batch_del_iterative(uri_str = None, bucket = None):
+    """ Streaming version of batch deletion (doesn't realize whole list in memory before deleting). 
+
+    Differences from subcmd_batch_del:
+      - Obeys cfg.max_delete, but if it is >1000, will delete files until we are sure that at least 
```

Comment:
That is not great at all to have such an inconsistency between batch_del and batch_del_iterative.
To sum up:
In one case, the "del" is cancelled if you have too much entries to remove.
In the other one, there is no file deleted if cfg.max_delete < 1000.
But, if > 1000,  a number of files rounded to the last thousand is deleted. It is not even that you have up to cfg.max_delete file deleted. 

I think that it would be a lot better to fallback to "subcmd_batch_del" if cfg.max_delete > 1000.